crates/bevy_animation/src/graph.rs

```diff
@@ -0,0 +1,446 @@
+    /// animation evaluation, the descendants of this blend node will have their
+    /// weights multiplied by the weight of the blend.
+    pub fn add_blend(&mut self, weight: f32, parent: AnimationNodeIndex) -> AnimationNodeIndex {
+        let node_index = self
+            .graph
+            .add_node(AnimationGraphNode { clip: None, weight });
+        self.graph.add_edge(parent, node_index, ());
+        node_index
+    }
+
+    /// Adds an edge from the edge `from` to `to`, making `to` a child of
+    /// `from`.
+    ///
+    /// The behavior is unspecified if adding this produces a cycle in the
+    /// graph.
+    pub fn add_edge(&mut self, from: NodeIndex, to: NodeIndex) {
+        self.graph.add_edge(from, to, ());
+    }
+
+    /// Removes an edge between `from` and `to` if it exists.
+    ///
+    /// Returns true if the edge was successfully removed or false if no such
+    /// edge existed.
+    pub fn remove_edge(&mut self, from: NodeIndex, to: NodeIndex) -> bool {
+        self.graph
+            .find_edge(from, to)
+            .map(|edge| self.graph.remove_edge(edge))
+            .is_some()
+    }
+
+    /// Returns the [`AnimationGraphNode`] associated with the given index.
+    ///
+    /// If no node with the given index exists, returns `None`.
+    pub fn get(&self, animation: AnimationNodeIndex) -> Option<&AnimationGraphNode> {
+        self.graph.node_weight(animation)
+    }
+
+    /// Returns a mutable reference to the [`AnimationGraphNode`] associated
+    /// with the given index.
+    ///
+    /// If no node with the given index exists, returns `None`.
+    pub fn get_mut(&mut self, animation: AnimationNodeIndex) -> Option<&mut AnimationGraphNode> {
+        self.graph.node_weight_mut(animation)
+    }
+
+    /// Returns an iterator over the [`AnimationGraphNode`]s in this graph.
+    pub fn nodes(&self) -> impl Iterator<Item = AnimationNodeIndex> {
+        self.graph.node_indices()
+    }
+
+    /// Performs a depth-first search on the animation graph.
+    pub(crate) fn dfs(
+        &self,
+    ) -> Dfs<AnimationNodeIndex, <Graph<AnimationGraphNode, ()> as Visitable>::Map> {
+        Dfs::new(&self.graph, self.root)
+    }
+
+    /// Serializes the animation graph to the given [`Write`]r in RON format.
+    pub fn save<W>(&self, writer: &mut W) -> Result<(), AnimationGraphLoadError>
+    where
+        W: Write,
+    {
+        let mut ron_serializer = ron::ser::Serializer::new(writer, None)?;
+        Ok(self.serialize(&mut ron_serializer)?)
+    }
+
+    /// A convenience method to serialize the animation graph to a file.
+    ///
+    /// This file can later be loaded with the [`AnimationGraphAssetLoader`] to
+    /// reconstruct the graph.
+    pub fn save_to<P>(&self, path: &P) -> Result<(), AnimationGraphLoadError>
+    where
+        P: AsRef<Path>,
+    {
+        let mut writer = BufWriter::new(File::create(path)?);
+        self.save(&mut writer)
+    }
```

Comment:
```suggestion
```

Not a fan of this helper. we should avoid APIs accessing files without bevy_assets

Reply:
I removed it.